diff --git a/lib/nproc.c b/lib/nproc.c
index ef86975..f09b429 100644
--- a/lib/nproc.c
+++ b/lib/nproc.c
@@ -62,6 +62,10 @@
 
 #include "minmax.h"
 
+#ifdef __MVS__
+#include <zos.h>
+#endif
+
 #define NPROC_MINIMUM 1
 
 /* Return the number of processors available to the current process, based
@@ -353,6 +357,10 @@ num_processors_available (enum nproc_query query)
   }
 #endif
 
+#if defined(__MVS__)
+  return __get_num_online_cpus();
+#endif
+
   return NPROC_MINIMUM;
 }
 
