diff --git a/lib/nproc.c b/lib/nproc.c
index b0c9514..79cc5bc 100644
--- a/lib/nproc.c
+++ b/lib/nproc.c
@@ -61,6 +61,10 @@
 
 #include "minmax.h"
 
+#ifdef __MVS__
+#include <zos.h>
+#endif
+
 #define ARRAY_SIZE(a) (sizeof (a) / sizeof ((a)[0]))
 
 #define NPROC_MINIMUM 1
@@ -354,6 +358,10 @@ num_processors_available (enum nproc_query query)
   }
 #endif
 
+#if defined(__MVS__)
+  return __get_num_online_cpus();
+#endif
+
   return NPROC_MINIMUM;
 }
 
