diff --git a/lib/posix_memalign.c b/lib/posix_memalign.c
index 7078af4..5aef7dc 100644
--- a/lib/posix_memalign.c
+++ b/lib/posix_memalign.c
@@ -26,6 +26,11 @@ int
 posix_memalign (void **memptr, size_t alignment, size_t size)
 #undef posix_memalign
 {
+/* FIXME: MVS does not have posix_memalign */
+#ifdef __MVS__
+    *memptr = malloc(size);
+    return *memptr;
+#else
   /* Round SIZE up to the next multiple of ALIGNMENT.
      However, treat a zero size as if it were ALIGNMENT.  */
   size_t aligned_size;
@@ -33,4 +38,5 @@ posix_memalign (void **memptr, size_t alignment, size_t size)
     return ENOMEM;
   aligned_size &= -alignment;
   return posix_memalign (memptr, alignment, aligned_size);
+#endif
 }
