diff --git a/src/copy.c b/src/copy.c
index b9fff03..4683399 100644
--- a/src/copy.c
+++ b/src/copy.c
@@ -1216,6 +1216,18 @@ handle_clone_fail (int dst_dirfd, char const *dst_relname,
   return true;
 }
 
+#ifdef __MVS__
+int __setfdccsid(int fd, int t_ccsid) 
+{
+  attrib_t attr;
+  memset(&attr, 0, sizeof(attr));
+  attr.att_filetagchg = 1;
+  attr.att_filetag.ft_txtflag = (t_ccsid >> 16);
+  attr.att_filetag.ft_ccsid = (t_ccsid & 0x0ffff);
+  return __fchattr(fd, &attr, sizeof(attr));
+}
+#endif
+
 
 /* Copy a regular file from SRC_NAME to DST_NAME aka DST_DIRFD+DST_RELNAME.
    If the source file contains holes, copies holes and blocks of zeros
@@ -1507,6 +1519,11 @@ copy_reg (char const *src_name, char const *dst_name,
       goto close_src_desc;
     }
 
+#ifdef __MVS__
+/* Copy MVS file tags */
+  __setfdccsid(dest_desc,  (src_sb->st_tag.ft_txtflag << 16) | src_sb->st_tag.ft_ccsid);
+#endif
+
   /* --attributes-only overrides --reflink.  */
   if (data_copy_required && x->reflink_mode)
     {
