diff --git a/src/copy.c b/src/copy.c
index b9fff03..554e433 100644
--- a/src/copy.c
+++ b/src/copy.c
@@ -113,6 +113,10 @@
 # define CAN_HARDLINK_SYMLINKS 0
 #endif
 
+#ifdef __MVS__
+# include "zos-io.h"
+#endif
+
 struct dir_list
 {
   struct dir_list *parent;
@@ -1507,6 +1511,11 @@ copy_reg (char const *src_name, char const *dst_name,
       goto close_src_desc;
     }
 
+#ifdef __MVS__
+/* Copy MVS file tags */
+  __setfdccsid(dest_desc,  (src_sb->st_tag.ft_txtflag << 16) | src_sb->st_tag.ft_ccsid);
+#endif
+
   /* --attributes-only overrides --reflink.  */
   if (data_copy_required && x->reflink_mode)
     {
