diff --git a/src/cp.c b/src/cp.c
index 317d667..f347ab5 100644
--- a/src/cp.c
+++ b/src/cp.c
@@ -607,6 +607,36 @@ make_dir_parents_private (char const *const_dir, size_t src_offset,
   return true;
 }
 
+#ifdef __MVS__
+#define MVS_COPY "/bin/cp"
+#define MVS_COPY_LEN (sizeof(MVS_COPY)-1)
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+static bool
+do_mvs_copy (int argc, char **argv)
+{
+
+  pid_t pid = fork();
+  if (pid == 0) {
+    argv[0] = MVS_COPY;
+    execvp(MVS_COPY, argv);
+  } else if (pid > 0) {
+    int wstatus;
+    pid_t waitchild = waitpid(pid, &wstatus, WUNTRACED | WCONTINUED);
+    if (waitchild == -1) {
+      perror("waitpid");
+      return false;
+    }
+    if (WIFEXITED(wstatus)) {
+      return WEXITSTATUS(wstatus) == 0 ? true : false;
+    } else if (WIFSIGNALED(wstatus)) {
+      fprintf(stderr, "killed by signal %d\n", WTERMSIG(wstatus));
+    }
+  }
+}
+#endif
+
 /* Scan the arguments, and copy each by calling copy.
    Return true if successful.  */
 
@@ -985,6 +1015,19 @@ main (int argc, char **argv)
   bool no_target_directory = false;
   char const *scontext = nullptr;
 
+#ifdef __MVS__
+  {
+    int i;
+
+    for (i=0; i<argc; ++i) {
+      if (argv[i][0] == '/' && argv[i][1] == '/') {
+        ok = do_mvs_copy(argc, argv);
+        main_exit (ok ? EXIT_SUCCESS : EXIT_FAILURE);
+      }
+    }
+  }
+#endif
+
   initialize_main (&argc, &argv);
   set_program_name (argv[0]);
   setlocale (LC_ALL, "");
@@ -1102,6 +1145,10 @@ main (int argc, char **argv)
           x.preserve_ownership = true;
           x.preserve_mode = true;
           x.preserve_timestamps = true;
+#ifdef __MVS__
+          x.preserve_xattr = true;
+          x.require_preserve_xattr = true;
+#endif
           x.require_preserve = true;
           break;
 
@@ -1266,7 +1313,7 @@ main (int argc, char **argv)
            _("failed to set default file creation context to %s"),
            quote (scontext));
 
-#if !USE_XATTR
+#if !USE_XATTR && !defined(__MVS__)
   if (x.require_preserve_xattr)
     error (EXIT_FAILURE, 0, _("cannot preserve extended attributes, cp is "
                               "built without xattr support"));
