diff --git a/src/digest.c b/src/digest.c
index 3d239da..22e5a4b 100644
--- a/src/digest.c
+++ b/src/digest.c
@@ -451,7 +451,7 @@ Print or check %s (%d-bit) checksums.\n\
 #endif
 #if !HASH_ALGO_SUM
 # if !HASH_ALGO_CKSUM
-      if (O_BINARY)
+      if (O_BINARY || __MVS__)
         fputs (_("\
   -b, --binary          read in binary mode (default unless reading tty stdin)\
 \n\
@@ -487,7 +487,7 @@ Print or check %s (%d-bit) checksums.\n\
 "), stdout);
 # endif
 # if !HASH_ALGO_CKSUM
-      if (O_BINARY)
+      if (O_BINARY || __MVS__)
         fputs (_("\
   -t, --text            read in text mode (default if reading tty stdin)\n\
 "), stdout);
@@ -974,7 +974,7 @@ digest_file (char const *filename, int *binary, unsigned char *bin_result,
     {
       have_read_stdin = true;
       fp = stdin;
-      if (O_BINARY && *binary)
+      if ((O_BINARY || __MVS__) && *binary)
         {
           if (*binary < 0)
             *binary = ! isatty (STDIN_FILENO);
@@ -984,7 +984,7 @@ digest_file (char const *filename, int *binary, unsigned char *bin_result,
     }
   else
     {
-      fp = fopen (filename, (O_BINARY && *binary ? "rb" : "r"));
+      fp = fopen (filename, ((O_BINARY || __MVS__) && *binary ? "rb" : "r"));
       if (fp == nullptr)
         {
           if (ignore_missing && errno == ENOENT)
@@ -997,6 +997,11 @@ digest_file (char const *filename, int *binary, unsigned char *bin_result,
         }
     }
 
+#ifdef __MVS__
+  if (*binary)
+    __disableautocvt(fileno(fp));
+#endif
+
   fadvise (fp, FADVISE_SEQUENTIAL);
 
 #if HASH_ALGO_CKSUM
@@ -1590,7 +1595,7 @@ main (int argc, char **argv)
      usage (EXIT_FAILURE);
    }
 
-  if (!O_BINARY && binary < 0)
+  if (!(O_BINARY || __MVS__) && binary < 0)
     binary = 0;
 
   char **operand_lim = argv + argc;
