diff --git a/src/stat.c b/src/stat.c
index dd86450..4c9aaf5 100644
--- a/src/stat.c
+++ b/src/stat.c
@@ -98,7 +98,14 @@
 # elif HAVE_STRUCT_STATFS_F_NAMEMAX
 #  define SB_F_NAMEMAX(S) ((S)->f_namemax)
 # endif
-# define STATFS statfs
+
+#ifdef __MVS__
+#include <sys/statfs.h>
+# define STATFS w_statfs
+#else
+ # define STATFS statfs
+#endif
+
 # if HAVE_OS_H /* BeOS */
 /* BeOS has a statvfs function, but it does not return sensible values
    for f_files, f_ffree and f_favail, and lacks f_type, f_basetype and
@@ -132,7 +139,11 @@ statfs (char const *filename, struct fs_info *buf)
 #  define STRUCT_STATXFS_F_FSID_IS_INTEGER true
 #  define STATFS_FRSIZE(S) ((S)->block_size)
 # else
+#ifndef __MVS__
 #  define STRUCT_STATVFS struct statfs
+#else
+#  define STRUCT_STATVFS struct w_statfs
+#endif
 #  define STRUCT_STATXFS_F_FSID_IS_INTEGER STRUCT_STATFS_F_FSID_IS_INTEGER
 #  if HAVE_STRUCT_STATFS_F_FRSIZE
 #   define STATFS_FRSIZE(S) ((S)->f_frsize)
@@ -257,8 +268,12 @@ NODISCARD
 static char const *
 human_fstype (STRUCT_STATVFS const *statfsbuf)
 {
-#ifdef STATXFS_FILE_SYSTEM_TYPE_MEMBER_NAME
+#ifdef STATXFS_FILE_SYSTEM_TYPE_MEMBER_NAME 
+#if defined __MVS__
+  return 0;
+#else
   return statfsbuf->STATXFS_FILE_SYSTEM_TYPE_MEMBER_NAME;
+#endif
 #else
   switch (statfsbuf->f_type)
     {
@@ -873,6 +888,9 @@ print_statfs (char *pformat, size_t prefix_len, MAYBE_UNUSED char mod, char m,
       {
 #if STRUCT_STATXFS_F_FSID_IS_INTEGER
         uintmax_t fsid = statfsbuf->f_fsid;
+        out_uint_x (pformat, prefix_len, fsid);
+#elif __MVS__
+        ;
 #else
         typedef unsigned int fsid_word;
         static_assert (alignof (STRUCT_STATVFS) % alignof (fsid_word) == 0);
@@ -890,8 +908,8 @@ print_statfs (char *pformat, size_t prefix_len, MAYBE_UNUSED char mod, char m,
             uintmax_t u = p[words - 1 - i];
             fsid |= u << (i * CHAR_BIT * sizeof *p);
           }
-#endif
         out_uint_x (pformat, prefix_len, fsid);
+#endif
       }
       break;
 
@@ -909,30 +927,44 @@ print_statfs (char *pformat, size_t prefix_len, MAYBE_UNUSED char mod, char m,
       out_string (pformat, prefix_len, human_fstype (statfsbuf));
       break;
     case 'b':
+#ifndef __MVS__
       out_int (pformat, prefix_len, statfsbuf->f_blocks);
+#endif
       break;
     case 'f':
+#ifndef __MVS__
       out_int (pformat, prefix_len, statfsbuf->f_bfree);
+#endif
       break;
     case 'a':
+#ifndef __MVS__
       out_int (pformat, prefix_len, statfsbuf->f_bavail);
+#endif
       break;
     case 's':
+#ifndef __MVS__
       out_uint (pformat, prefix_len, statfsbuf->f_bsize);
+#endif
       break;
     case 'S':
       {
+#ifndef __MVS__
         uintmax_t frsize = STATFS_FRSIZE (statfsbuf);
         if (! frsize)
           frsize = statfsbuf->f_bsize;
         out_uint (pformat, prefix_len, frsize);
+#endif
       }
       break;
     case 'c':
+#ifndef __MVS__
       out_uint (pformat, prefix_len, statfsbuf->f_files);
+#endif
       break;
     case 'd':
+#ifndef __MVS__
       out_int (pformat, prefix_len, statfsbuf->f_ffree);
+#endif
       break;
     default:
       fputc ('?', stdout);
@@ -1267,12 +1299,14 @@ do_statfs (char const *filename, char const *format)
       return false;
     }
 
+#ifndef __MVS__
   if (STATFS (filename, &statfsbuf) != 0)
     {
       error (0, errno, _("cannot read file system information for %s"),
              quoteaf (filename));
       return false;
     }
+#endif
 
   bool fail = print_it (format, -1, filename, print_statfs, &statfsbuf);
   return ! fail;
