diff --git a/src/cp.c b/src/cp.c
index 0f44b35..b84f163 100644
--- a/src/cp.c
+++ b/src/cp.c
@@ -565,6 +565,15 @@ make_dir_parents_private (char const *const_dir, size_t src_offset,
   return true;
 }
 
+#ifdef __MVS__
+static bool
+do_mvs_copy (int argc, char **argv) 
+{
+  printf("mvs copy to be done\n");
+  return true;
+}
+#endif
+
 /* Scan the arguments, and copy each by calling copy.
    Return true if successful.  */
 
@@ -1212,6 +1221,20 @@ main (int argc, char **argv)
 
   hash_init ();
 
+#ifdef __MVS__
+  {
+    int i;
+    int nfiles=argc-optind;
+    char** files = argv+optind;
+
+    for (i=0; i<nfiles; ++i) {
+      if (files[i][0] == '/' && files[i][1] == '/') {
+        ok = do_mvs_copy(argc, argv);
+        main_exit (ok ? EXIT_SUCCESS : EXIT_FAILURE);
+      }
+    }
+  }
+#endif
   ok = do_copy (argc - optind, argv + optind,
                 target_directory, no_target_directory, &x);
 
